crates/xtask/src/changelog.rs

```diff
@@ -0,0 +1,377 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use anyhow::{ensure, Context, Result};
+use clap::ValueEnum;
+
+#[derive(clap::ValueEnum, Clone, Debug)]
+pub enum ChangelogReleaseType {
+    Major,
+    Minor,
+    Patch,
+}
+
+/// Changelog file data structure.
+#[derive(Debug, Default, PartialEq)]
+struct Changelog {
+    releases: Vec<ChangelogRelease>,
+
+    increment_count: u32,
+}
+
+impl Changelog {
+    fn push_release(&mut self, release: ChangelogRelease) {
+        self.releases.push(release);
+    }
+
+    fn set_increment_count(&mut self, num: u32) {
+        self.increment_count = num;
+    }
+}
+
+/// Each release of a Changelog, containing a version string and optional major, minor, patch change
+/// descriptions.
+#[derive(Debug, PartialEq)]
+struct ChangelogRelease {
+    version: String,
+
+    major: Option<Vec<String>>,
+    minor: Option<Vec<String>>,
+    patch: Option<Vec<String>>,
+}
+
+impl ChangelogRelease {
+    fn from(version: String) -> Self {
+        ChangelogRelease { version, major: None, minor: None, patch: None }
+    }
+
+    fn set_major(&mut self, contents: Vec<String>) {
+        self.major = Some(contents);
+    }
+
+    fn set_minor(&mut self, contents: Vec<String>) {
+        self.minor = Some(contents);
+    }
+
+    fn set_patch(&mut self, contents: Vec<String>) {
+        self.patch = Some(contents);
+    }
+}
+
+pub fn execute(crate_path: String, _version: ChangelogReleaseType, _message: String) -> Result<()> {
```

Comment:
We don't need to take ownership of arguments, except `ReleaseType` which should be `Copy` anyway.

For paths, we can either take `&str`, or even better `&Path` (or `&impl AsRef<Path>`). But given the rest of xtask uses `String` instead of `PathBuf` for file flags, then we should use `&str` here.